### ISSUE TITLE
rustfmt fails to format matches! call with enum wildcard destructuring and joined patterns

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rustfmt#4512]
**Input**

```rust
enum X {
    A(String),
    B(u32),
}

fn fun(a: X, b: X) -> bool {
    matches!((a,b), (X::A(_),

X::B(_))|
(X::B(_),


 X::A(_)))
}
```

**Output**

```rust
enum X {
    A(String),
    B(u32),
}

fn fun(a: X, b: X) -> bool {
    matches!((a,b), (X::A(_),

X::B(_))|
(X::B(_),


 X::A(_)))
}
```

**Expected output**
something more like:
```rust
enum X {
    A(String),
    B(u32),
}

fn fun(a: X, b: X) -> bool {
    matches!((a ,b),
           (X::A(_), X::B(_))
           | (X::B(_), X::A(_)))
}
```

**Meta**
`rustfmt 1.4.22-nightly (97d0301 2020-10-04)`

- From where did you install rustfmt?: rustup>

**Comment by calebcartwright** [comment]
This doesn't really have anything to do with tuples, and despite the semantics, nor with matches either. The issue is really with the failure to parse that second arg which occurs when there's both a semantic enum wildcard destruct and a `|` operator from a pattern join. 

Simple repro with exaggerated spacing for effect to show rustfmt bailing and reverting to the original formatting:
```rust
fn f() {
    matches!(c,
    
    A(_)| 
    
    B(_)
);
            }
```

**Comment by calebcartwright** [comment]
This has been fixed, presumably via the various upstream work in rustc around the token streams/macro handling. Improvements are already available on master, and included in the upcoming v1.4.28 release.

Closing accordingly

**Comment by matthiaskrgr** [comment]
that's great to hear! :)
